### PR TITLE
Fix usage of 'distinct' combinator with 'state' combinator

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix usage of `-Distinct` combinator with `-State` combinator in aggregate functions.

### CODE REVIEW

**Comment by CurtizJ** [comment]
If function is marked with `State` combinator, arena will be added [here](https://github.com/ClickHouse/ClickHouse/blob/80e6ad7aae080ad9d3ab3d49a67214f8d713c11a/src/Interpreters/Aggregator.cpp#L1209-L1225) to `ColumnAggregateFunction`, otherwise it will be destroyed after aggregating or merging blocks and `use-after-free` is possible. It reproduces only with `distributed_aggregation_memory_efficient`, but I don't know, why it works in other cases.

File: src/AggregateFunctions/AggregateFunctionDistinct.h [hunk 1]
```diff
@@ -236,6 +236,11 @@ class AggregateFunctionDistinct : public IAggregateFunctionDataHelper<Data, Aggr
         return true;
     }
 
+    bool isState() const override
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/AggregateFunctionDistinct.h:239]: Ok. It means we can have both `aggDistinctState` and `aggStateDistinct` that will act similar?

**Review by alexey-milovidov — APPROVED** [review]

File: src/AggregateFunctions/AggregateFunctionDistinct.h (hunk 1 above)
Comment (CurtizJ) [reply src/AggregateFunctions/AggregateFunctionDistinct.h:239]: Yes.

**Comment by alexey-milovidov** [comment]
Integration tests - spurious failure.

**Pull request merged**